File: src/DcricleScan/src/api/PullEssenceMessage.ts
import {Send} from "./api";
import {ClassArray} from "../3rdparty/ts-json";
import {Encrypted} from "../helper/EncryptorBuilderFactory";
import {Type} from "../helper/Message";
import {DBMessage} from "../db/DBMessage";
import {GetDScan} from "../db/db";
import {getUs} from "../DIDBrowser";

export class PullEssenceMessageRequest {
  public chatId:string = "";
  public end:number = -1;
  public limit:number = 10;
}

export class PullEssenceMessageItem implements DBMessage.Document{
  public msgId:string = '';
  public chatId:string = '';
  public realChatId:string = "";
  public seq:number = -1;
  public content:string = '';
  public encrypted:Encrypted = Encrypted.AES;
  public creatorUid:string = '';
  public createTime:number = 0;
  public msgType:Type = Type.Text;
}

class PullEssenceMessageResponse {
  public items: ClassArray<PullEssenceMessageItem> = new ClassArray<PullEssenceMessageItem>(PullEssenceMessageItem)
}
export async function PullEssenceMessage(request: PullEssenceMessageRequest):Promise<[PullEssenceMessageResponse, Error|null]> {
  const [ret, err] = await Send("/browser/PullEssenceMessage", request, PullEssenceMessageResponse);
  if (err) {
    return [new PullEssenceMessageResponse(), err]
  }
  if (ret.items.length <= 0) {
    return [new PullEssenceMessageResponse(), null]
  }
  for(let i = 0; i < ret.items.length; i++) {
    const item = ret.items[i];
    item.realChatId = request.chatId
  }
  await DBMessage.Insert(GetDScan(), ...ret.items);
  await getUs().nc.post(new DBMessage.ChangedEvent([request.chatId]))
  return [ret, err]
}

File: src/DcircleChain/node/chain/api/PullEssenceMessage.go
package api

import (
	"context"
	"dcircleserver/src/im/db"
	"dcircleserver/src/im/db/message"
	"fmt"
	"github.com/xpwu/go-log/log"
	"math"
)

type PullEssenceMessageRequest struct {
	ChatId string `json:"chatId"`
	End    int64  `json:"end"` //传-1时 为从头再拉
	Limit  int64  `json:"limit"`
}

type PullEssenceMessageItem struct {
	MsgId      string              `json:"msgId"`
	ChatId     string              `json:"chatId"`
	Seq        int64               `json:"seq"`
	Content    string              `json:"content"`
	Encrypted  db.MessageEncrypted `json:"encrypted"`
	CreatorUid string              `json:"creatorUid"`
	CreateTime int64               `json:"createTime"`
	MsgType    message.Type        `json:"msgType"`
}

type PullEssenceMessageResponse struct {
	Items []PullEssenceMessageItem `json:"items"`
}

func (s *suite) APIPullEssenceMessage(ctx context.Context, request *PullEssenceMessageRequest) *PullEssenceMessageResponse {
	_, logger := log.WithCtx(ctx)
	logger.PushPrefix("PullEssenceMessage")
	logger.Debug(fmt.Sprintf("req: %v", request))
	res := &PullEssenceMessageResponse{Items: []PullEssenceMessageItem{}}
	if len(request.ChatId) == 0 {
		logger.Debug("ChatId is nil")
		return res
	}

	limit := request.Limit
	const maxLimit = 50
	if limit > maxLimit {
		limit = maxLimit
	}

	if limit <= 0 {
		limit = maxLimit
	}

	const maxSeq = math.MaxInt64
	end := request.End
	if end < 0 {
		end = maxSeq
	}

	esChat, err := db.NewChat(ctx).FindEssenceChatId(request.ChatId)
	if err != nil {
		logger.Error(err)
		s.Request.Terminate(err)
	}

	if len(esChat.ChatId) == 0 {
		logger.Debug("没有找到精华会话，chatId:", request.ChatId)
		return res
	}

	messageList, err := PullDownEssenceMessage(ctx, esChat.ChatId, end, limit)
	if err != nil {
		logger.Error(err)
		s.Request.Terminate(err)
	}

	for _, doc := range messageList {
		res.Items = append(res.Items, PullEssenceMessageItem{
			MsgId:      doc.MsgId,
			ChatId:     doc.ChatId,
			Seq:        doc.Seq,
			Content:    doc.Content,
			Encrypted:  doc.Encrypted,
			CreatorUid: doc.CreatorUid,
			CreateTime: doc.CreateTime,
			MsgType:    doc.MsgType,
		})
	}

	return res
}

func PullDownEssenceMessage(ctx context.Context, chatId string, end int64, limit int64) (msgList []*db.MessageDocument, err error) {
	ctx, logger := log.WithCtx(ctx)
	logger.PushPrefix("PullDownEssenceMessage")
	msgList = []*db.MessageDocument{}
	if limit <= 0 {
		panic(fmt.Sprintf("PullDownEssenceMessage limit:%v is err", limit))
	}

	if end <= 1 {
		logger.Debug("end ==1 直接返回")
		return
	}

	noAuthMap, err := db.NewAuth(ctx).GetNoAuthedBetweenSeq(chatId, []string{db.AllUid}, 0, end)
	if err != nil {
		logger.Error(err)
		return
	}

	var docs []*db.MessageDocument
	for {
		if end <= 1 {
			break
		}
		docs, err = db.NewMessage(ctx).ReadToEndSeq(chatId, end, limit)
		if err != nil {
			logger.Error(err)
			return
		}

		for _, doc := range docs {
			if _, ok := noAuthMap[doc.Seq]; ok {
				continue
			}
			msgList = append(msgList, doc)
		}

		if len(msgList) >= int(limit) {
			break
		}

		if len(docs) < int(limit) {
			break
		}
		end = docs[len(docs)-1].Seq
	}

	return
}
